File: backend/app/schemas/company_profile.py
from pydantic import BaseModel


class CompanyProfileCreate(BaseModel):
    user_id: str
    company_name: str | None = None
    entity_type: str | None = None
    incorporation_date: str | None = None
    industry_vertical: str | None = None
    industry_sub_sector: str | None = None
    business_model: str | None = None
    revenue_model: str | None = None
    avg_deal_size: float | None = None
    avg_contract_length_months: int | None = None
    funding_stage: str | None = None
    last_raise_amount: float | None = None
    total_capital_raised: float | None = None
    last_raise_date: str | None = None
    cash_reserves: float | None = None
    monthly_burn_rate: float | None = None
    monthly_revenue: float | None = None
    runway_target_months: int | None = None
    cac: float | None = None
    ltv: float | None = None
    payback_period_months: int | None = None
    gross_margin_target: float | None = None
    net_margin_target: float | None = None
    customer_type: str | None = None
    customer_segment: str | None = None
    avg_contract_value: float | None = None
    monthly_churn_rate: float | None = None
    team_size: int | None = None
    engineering_headcount: int | None = None
    sales_headcount: int | None = None
    ops_headcount: int | None = None
    contractor_count: int | None = None
    next_fundraise_date: str | None = None
    profitability_target_date: str | None = None
    revenue_target_3m: float | None = None
    revenue_target_6m: float | None = None
    revenue_target_12m: float | None = None
    exit_strategy: str | None = None
    operating_states: str | None = None
    has_gst_registration: bool = False
    has_tan_registration: bool = False
    gstin: str | None = None
    pan: str | None = None
    auditor_appointed: bool = False
    statutory_audit_required: bool = False
    current_fy: str | None = None
    profile_completeness: int = 0
    last_reviewed_at: str | None = None


class CompanyProfileUpdate(BaseModel):
    company_name: str | None = None
    entity_type: str | None = None
    incorporation_date: str | None = None
    industry_vertical: str | None = None
    industry_sub_sector: str | None = None
    business_model: str | None = None
    revenue_model: str | None = None
    avg_deal_size: float | None = None
    avg_contract_length_months: int | None = None
    funding_stage: str | None = None
    last_raise_amount: float | None = None
    total_capital_raised: float | None = None
    last_raise_date: str | None = None
    cash_reserves: float | None = None
    monthly_burn_rate: float | None = None
    monthly_revenue: float | None = None
    runway_target_months: int | None = None
    cac: float | None = None
    ltv: float | None = None
    payback_period_months: int | None = None
    gross_margin_target: float | None = None
    net_margin_target: float | None = None
    customer_type: str | None = None
    customer_segment: str | None = None
    avg_contract_value: float | None = None
    monthly_churn_rate: float | None = None
    team_size: int | None = None
    engineering_headcount: int | None = None
    sales_headcount: int | None = None
    ops_headcount: int | None = None
    contractor_count: int | None = None
    next_fundraise_date: str | None = None
    profitability_target_date: str | None = None
    revenue_target_3m: float | None = None
    revenue_target_6m: float | None = None
    revenue_target_12m: float | None = None
    exit_strategy: str | None = None
    operating_states: str | None = None
    has_gst_registration: bool | None = None
    has_tan_registration: bool | None = None
    gstin: str | None = None
    pan: str | None = None
    auditor_appointed: bool | None = None
    statutory_audit_required: bool | None = None
    current_fy: str | None = None
    profile_completeness: int | None = None
    last_reviewed_at: str | None = None


class CompanyProfileResponse(BaseModel):
    id: str
    user_id: str
    company_name: str | None
    entity_type: str | None
    incorporation_date: str | None
    industry_vertical: str | None
    industry_sub_sector: str | None
    business_model: str | None
    revenue_model: str | None
    avg_deal_size: float | None
    avg_contract_length_months: int | None
    funding_stage: str | None
    last_raise_amount: float | None
    total_capital_raised: float | None
    last_raise_date: str | None
    cash_reserves: float | None
    monthly_burn_rate: float | None
    monthly_revenue: float | None
    runway_target_months: int | None
    cac: float | None
    ltv: float | None
    payback_period_months: int | None
    gross_margin_target: float | None
    net_margin_target: float | None
    customer_type: str | None
    customer_segment: str | None
    avg_contract_value: float | None
    monthly_churn_rate: float | None
    team_size: int | None
    engineering_headcount: int | None
    sales_headcount: int | None
    ops_headcount: int | None
    contractor_count: int | None
    next_fundraise_date: str | None
    profitability_target_date: str | None
    revenue_target_3m: float | None
    revenue_target_6m: float | None
    revenue_target_12m: float | None
    exit_strategy: str | None
    operating_states: str | None
    has_gst_registration: bool
    has_tan_registration: bool
    gstin: str | None
    pan: str | None
    auditor_appointed: bool
    statutory_audit_required: bool
    current_fy: str | None
    profile_completeness: int
    last_reviewed_at: str | None
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True

File: src/types/company-profile.ts
/**
 * @file Company Profile Types
 * @description Enriched company profile for contextual financial intelligence
 */

// Entity types for Indian companies
export type EntityType = 'pvt_ltd' | 'llp' | 'opc' | 'partnership' | 'sole_proprietor';

export type BusinessModel = 'saas' | 'marketplace' | 'd2c' | 'services' | 'hardware' | 'hybrid';

export type RevenueModel = 'subscription' | 'transactional' | 'one_time' | 'usage_based' | 'hybrid';

export type FundingStage = 'bootstrapped' | 'pre_seed' | 'seed' | 'series_a' | 'series_b' | 'series_c_plus';

export type CustomerType = 'b2b' | 'b2c' | 'b2b2c';

export type CustomerSegment = 'enterprise' | 'mid_market' | 'smb' | 'consumer';

export type ExitStrategy = 'ipo' | 'acquisition' | 'lifestyle' | 'undecided';

export type IndustryVertical = 'fintech' | 'healthtech' | 'edtech' | 'ecommerce' | 'saas' | 'marketplace' | 'deeptech' | 'agritech' | 'cleantech' | 'logistics' | 'media' | 'gaming' | 'other';

export type StartupStage = 'pre-revenue' | 'early' | 'scaling' | 'growth';

export interface CompanyProfile {
  id?: string;
  userId?: string;

  // Company Basics
  companyName: string;
  entityType?: EntityType;
  incorporationDate?: string;
  industryVertical?: IndustryVertical;
  industrySubSector?: string;

  // Business Model
  businessModel?: BusinessModel;
  revenueModel?: RevenueModel;
  avgDealSize?: number;
  avgContractLengthMonths?: number;

  // Funding
  fundingStage?: FundingStage;
  lastRaiseAmount?: number;
  totalCapitalRaised?: number;
  lastRaiseDate?: string;

  // Cash & Burn
  cashReserves?: number;
  monthlyBurnRate?: number;
  monthlyRevenue?: number;
  runwayTargetMonths?: number;

  // Unit Economics
  cac?: number;
  ltv?: number;
  paybackPeriodMonths?: number;
  grossMarginTarget?: number;
  netMarginTarget?: number;

  // Customers
  customerType?: CustomerType;
  customerSegment?: CustomerSegment;
  avgContractValue?: number;
  monthlyChurnRate?: number;

  // Team
  teamSize?: number;
  engineeringHeadcount?: number;
  salesHeadcount?: number;
  opsHeadcount?: number;
  contractorCount?: number;

  // Financial Goals
  nextFundraiseDate?: string;
  profitabilityTargetDate?: string;
  revenueTarget3m?: number;
  revenueTarget6m?: number;
  revenueTarget12m?: number;
  exitStrategy?: ExitStrategy;

  // Compliance
  operatingStates: string[];
  hasGSTRegistration: boolean;
  hasTANRegistration: boolean;
  gstin?: string;
  pan?: string;
  auditorAppointed?: boolean;
  statutoryAuditRequired?: boolean;
  currentFY: string;

  // Meta
  profileCompleteness?: number;
  lastReviewedAt?: string;

  // Legacy compatibility (maps to old StartupProfile)
  stage: StartupStage;
}

// Helper to compute profile completeness
export function computeProfileCompleteness(profile: CompanyProfile): number {
  const fields = [
    profile.companyName,
    profile.entityType,
    profile.industryVertical,
    profile.businessModel,
    profile.revenueModel,
    profile.fundingStage,
    profile.monthlyRevenue,
    profile.cashReserves,
    profile.monthlyBurnRate,
    profile.customerType,
    profile.customerSegment,
    profile.teamSize,
    profile.operatingStates?.length > 0,
    profile.hasGSTRegistration !== undefined,
    profile.hasTANRegistration !== undefined,
    profile.currentFY,
    profile.exitStrategy,
    profile.cac,
    profile.ltv,
    profile.grossMarginTarget,
  ];
  const filled = fields.filter(Boolean).length;
  return Math.round((filled / fields.length) * 100);
}

// Convert enriched profile to legacy StartupProfile format for backward compatibility
export function toLegacyProfile(profile: CompanyProfile): {
  name: string;
  stage: StartupStage;
  monthlyRevenue: number;
  teamSize: number;
  operatingStates: string[];
  hasGSTRegistration: boolean;
  hasTANRegistration: boolean;
  incorporationDate?: string;
  sector?: string;
  currentFY: string;
} {
  return {
    name: profile.companyName,
    stage: profile.stage,
    monthlyRevenue: profile.monthlyRevenue || 0,
    teamSize: profile.teamSize || 0,
    operatingStates: profile.operatingStates || [],
    hasGSTRegistration: profile.hasGSTRegistration,
    hasTANRegistration: profile.hasTANRegistration,
    incorporationDate: profile.incorporationDate,
    sector: profile.industryVertical,
    currentFY: profile.currentFY,
  };
}

// Default empty profile
export const DEFAULT_COMPANY_PROFILE: CompanyProfile = {
  companyName: '',
  operatingStates: [],
  hasGSTRegistration: false,
  hasTANRegistration: false,
  currentFY: '2025-26',
  stage: 'early',
};
